subsys/net/lib/ocpp/ocpp_j.c

```diff
@@ -0,0 +1,883 @@
+		msg.unknown_key = key;
+	}
+
+	struct json_obj_descr keyval_descr[] = {
+		JSON_OBJ_DESCR_PRIM_NAMED(struct json_ocpp_key_val, "key",
+					  key, JSON_TOK_STRING),
+		JSON_OBJ_DESCR_PRIM_NAMED(struct json_ocpp_key_val, "readonly",
+					  readonly, JSON_TOK_NUMBER),
+		JSON_OBJ_DESCR_PRIM_NAMED(struct json_ocpp_key_val, "value",
+					  value, JSON_TOK_STRING),
+	};
+
+	struct json_obj_descr config_descr[GET_CFG_MAX_FIELDS] = {
+		JSON_OBJ_DESCR_OBJ_ARRAY_NAMED(struct json_ocpp_getconfig_msg, "configurationKey",
+					       configuration_key, 1, configuration_key_len,
+					       keyval_descr, ARRAY_SIZE(keyval_descr))
+	};
+
+	if (!val) {
+		config_descr[0] = (struct json_obj_descr)
+		JSON_OBJ_DESCR_PRIM_NAMED(struct json_ocpp_getconfig_msg, "unknownKey",
+					  unknown_key, JSON_TOK_STRING);
+	}
+
+	ret = json_obj_encode_buf(config_descr, GET_CFG_MAX_FIELDS, &msg,
+				  tmp_buf, sizeof(tmp_buf));
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = frame_rpc_call_res(buf, len, uid, tmp_buf);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int frame_status_resp_msg(char *buf, int len, char *res, char *uid)
+{
+	int ret = 0;
+	char tmp_buf[JSON_MSG_BUF_128];
+
+	struct json_obj_descr descr[] = {
+		JSON_OBJ_DESCR_PRIM_NAMED(struct json_common_payload_field_str, "status",
+					  val1, JSON_TOK_STRING),
+	};
+
+	struct json_common_payload_field_str msg = {
+		.val1 = res,
+	};
+
+	ret = json_obj_encode_buf(descr, ARRAY_SIZE(descr), &msg,
+				  tmp_buf, sizeof(tmp_buf));
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = frame_rpc_call_res(buf, len, uid, tmp_buf);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+/* parse msg from server */
+int parse_rpc_msg(char *msg, int msglen, char *uid, int uidlen,
+		  int *pdu, bool *is_rsp)
+{
+	int ret = 0;
+	char local_buf[JSON_MSG_BUF_512];
+	char action[JSON_MSG_BUF_128];
+	char *token;
+	int rpc_id = -1;
+
+	if (!msg || !uid || !pdu || !is_rsp) {
+		return -EINVAL;
+	}
+
+	memcpy(local_buf, msg + 1, sizeof(local_buf) - 1);
+	local_buf[sizeof(local_buf) - 1] = '\0';
+
+	token = strtok(local_buf, ",");
+	if (!token) {
+		return -EINVAL;
+	}
+
+	rpc_id = *token - '0';
+
+	token = strtok(NULL, ",");
+	if (!token) {
+		return -EINVAL;
+	}
+
+	ret = extract_string_field(uid, uidlen, token);
+	if (ret < 0) {
+		return ret;
+	}
+
+	switch (rpc_id + '0') {
+	case OCPP_WAMP_RPC_REQ:
+		token = strtok(NULL, ",");
+		if (!token) {
+			return -EINVAL;
+		}
+
+		ret = extract_string_field(action, sizeof(action), token);
+		if (ret < 0) {
+			return ret;
+		}
+		*pdu = ocpp_find_pdu_from_literal(action);
+		/* fall through */
+		__fallthrough;
+
+	case OCPP_WAMP_RPC_RESP:
+		*is_rsp = rpc_id - 2;
+		ret = extract_payload(msg, msglen);
+		if (ret < 0) {
+			return ret;
+		}
+		break;
+
+	case OCPP_WAMP_RPC_ERR:
+		/* fall through */
+		__fallthrough;
+
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int parse_idtag_info(char *json, struct ocpp_idtag_info *idtag_info)
+{
+	int ret = 0;
```

Comment:
and here too, could you check all of these and remove unnecessary initializations